feat(migrations): Add state migrations

damassi opened the pull request:
This PR resolves [MOPLAT-478]

### Description

This adds state migrations to Folio, which behave identically to how we do it in Eigen. Along the way we found a chance to simplify the implementation however, and rely exclusively on easy-peasy's persistence implementation, versus needing to roll our own as we did in Eigen. 

Migrations live in `app/system/migrations.ts`. 

Still need to do a little cleanup and add some tests/docs, but this is ready for review. 

Follow-up is moving this new implementation back over to Eigen and remove all of the hand-rolled persistence code. 

**To Test**:
- Via flipper, install and enable `flipper-async-storage-advanced`
- Start energy and add a new migration
- Go into the flipper async storage pane and change `version` to something lower, hit update button
- Reload Energy in the simulator 

cc @artsy/mobile-platform 

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md


[MOPLAT-478]: https://artsyproduct.atlassian.net/browse/MOPLAT-478?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

damassi commented on package.json:
```diff
@@ -100,6 +100,7 @@
     "react-spring": "8.0.23",
     "relay-runtime": "^14.1.0",
     "relay-test-utils": "^14.1.0",
+    "rn-flipper-async-storage-advanced": "^1.0.4",
```
Gonna remove this

pvinis replied:
its not super useful, but i guess its kinda nice to have. though its one of those things that if it is ever bothering/blocking anything, it would go away in an instant.

damassi replied:
Moved to dev dep

damassi commented on src/app/App.tsx:
```diff
@@ -8,6 +9,7 @@ ignoreLogs()
 
 export const App = () => (
   <Providers>
+    <FlipperAsyncStorage />
```
And also gonna remove this, just for testing

pvinis replied:
ah, i remember temporarily adding this like
```
<>
<FlipperThing/>

<Providers>
<Main/>
</Prov>
<>
```
but like.. same. annoying that its a component

damassi replied:
This is another PR I was thinking about, removing this component requirement. I mean, even if it were a hook or something that'd be nicer than embedding a component.

damassi commented on src/app/Navigation.tsx:
```diff
@@ -45,7 +45,7 @@ type Screens = {
 
 export const Main = () => {
-  const isRehydrated = useStoreRehydrated()
+  const isDoneBooting = useSystemIsDoneBooting()
```
We needed to wrap `easy-peasy`'s hydration hook to ensure that we were done migrating. At that point the app is safe to boot.

pvinis replied:
what do you think about making a patch to add a `/** @deprecated at Artsy we use useSystemIsDoneBooting(). */` or something? 🤔 so if someone uses it accidentally, they get a strikethrough that might help them figure out the right thing to use.

damassi replied:
I'm going to really try to pair back patches unless strictly needed, even if they're a comment, across the board. This is something that will only be used once, so not too worried. But even if it weren't, rolling with the `system` concept for apps is something that I've found works really well; if there's a third party lib requirement that interfaces with this level of the stack, roll it into a general "system" init function. Everything goes there, and it becomes a sort of cultural understanding.

damassi commented on src/app/system/store/Models/GlobalStoreModel.ts:
```diff
@@ -40,10 +52,22 @@ export const getGlobalStoreModel = (): GlobalStoreModel => ({
 
+  // Migrates the locally persisted state to the latest version
+  performMigrations: action((state) => {
+    migrateState({ state })
```
This is the secret sauce in the PR. Rather than relying on an outside-the-system state / persistence update mechanism operating on JSON objects, just thread it right through easy-peasy's own state update model via an action, which simultaneously updates the persisted async storage in the background.

MounirDhahri replied:
Nice trick 👏

damassi commented on src/app/system/store/migrations.ts:
```diff
@@ -0,0 +1,51 @@
+  DummyMigration2: 2,
+  DummyMigration3: 3,
+  DummyMigration4: 4,
+  DummyMigration5: 5,
```
Will clear this stuff out on Monday but would like you to test / verify things are working good @pvinis 🙏

damassi commented on src/app/system/store/persistence/utils/migrateState.ts:
```diff
@@ -0,0 +1,30 @@
+      throw new Error("No migrator found for app version " + nextVersion)
+    }
+
+    migrator(state)
```
pass immerized state all the way through to the migrator function; no need for intermediate state representations anymore

damassi commented on src/app/system/store/persistence/migrateState.ts:
```diff
@@ -0,0 +1,30 @@
+      throw new Error("No migrator found for app version " + nextVersion)
+    }
+
+    migrator(state)
```
Pass immerized state all the way through to the migrator function from the action; no need for intermediate state representations anymore

gkartalis reviewed (approved):
Looking great! Only one thing, it would be nice to have a similar doc to the one we have in eigen for [migrations](https://github.com/artsy/eigen/blob/main/docs/adding_state_migrations.md)

MounirDhahri reviewed (approved):
Nice PR! thanks for assigning me

pvinis commented on src/app/system/store/persistence/migrateState.ts:
```diff
@@ -0,0 +1,30 @@
+import { CURRENT_APP_VERSION, energyAppMigrations, Migrations } from "app/system/store/migrations"
+
+export function migrateState<State extends { version: number }>({
```
you mean console log? if yes, then sure

MounirDhahri replied:
yes, that's what I had in mind

pvinis replied:
yep. and actually, it might be a good thing to send to sentry or segment. just a little event maybe 🤔?

MounirDhahri replied:
+++

pvinis commented on src/app/system/store/Models/GlobalStoreModel.ts:
```diff
@@ -20,6 +22,14 @@ interface GlobalStoreStateModel {
+  performMigrations: Action<this>
+  sessionState: {
+    isDonePerformingMigrations: boolean
+    setIsDonePerformingMigrations: Action<GlobalStoreStateModel>
```
where is this called? i cannot find it in this PR 🤔

pvinis reviewed (approved):
nice! thanks for taking this over here.

pvinis commented:
oh actually, one thing that is missing here from my branches are the tests. could you add the migration.tests.tsx back here?

damassi commented:
Yup @pvinis mentioned in the description. Still need to add the tests. Re `isDonePerformingMigrations`, that was something that i refactored late in the day. It will now be assigned in the `performMigrations()` action, no need for a second action for this. Need to update.

damassi commented:
Ok! Addressed feedback and merging this. 

At some point if my state migrations PR to easy peasy is accepted i'll loop back around and strip out all of this unnecessary baggage, and also update Eigen. 🤞 that they merge it.
